File: java/map_sum_pair_2.java
class MapSum {
    class TrieNode{
        TrieNode[] children = new TrieNode[26];
        int weight = 0;
    }
    TrieNode root;
    public MapSum() {
        root = new TrieNode();
    }
    
    public void insert(String key, int val) {
        TrieNode curr = root;
        for (int i = 0; i < key.length(); i++) {
            char c = key.charAt(i);
            if(curr.children[c-'a'] == null) {
                curr.children[c-'a'] = new TrieNode();
            }  
            curr = curr.children[c-'a'];
        }
        curr.weight = val;
    }
    
    public int sum(String prefix) {
        TrieNode curr = root;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            if(curr.children[c-'a'] == null) {
                return 0;   
            }
            curr = curr.children[c-'a'];
        }
        return dfs(curr);
    }
    
    private int dfs(TrieNode node){
        int sum = 0;
        for (int i = 0; i < 26; i++) {
            if (node.children[i] != null){
                sum += dfs(node.children[i]);
            }
        }
        return sum + node.weight;
    }
}

/**
 * Your MapSum object will be instantiated and called as such:
 * MapSum obj = new MapSum();
 * obj.insert(key,val);
 * int param_2 = obj.sum(prefix);
 */

File: js/map_sum_pair_2.js
var MapSum = function () {
    this.trie = {};
  };
  
  /** 
   * @param {string} key 
   * @param {number} val
   * @return {void}
   */
  MapSum.prototype.insert = function (key, val) {
    let t = this.trie;
    for (let i = 0; i < key.length; i++) {
      let char = key[i];
      if (!t[char]) {
          t[char] = {};
      }
      t = t[char];
    }
    t['value'] = val;
      console.log("node ", t);
    return null;
  };
  
  /** 
   * @param {string} prefix
   * @return {number}
   */
  MapSum.prototype.sum = function (prefix) {
    let node = this.trie
    for (let i = 0; i < prefix.length; i++) {
        let char = prefix[i];
        node = node[char];
        if (!node) {
            return 0;
        }
    }
    return dfs(node);
      
    function dfs(node) {
        let sum = 0;
        for (let k in node) {
        if (k === 'value') {
            sum += node[k];
        }
        else {
            sum += dfs(node[k]);
        }
      }
      return sum;
    }
  };